### PR TITLE
Add huge mushrooms and simple tree generator

### PR DESCRIPTION
**Pull request by jimmikaelkael**
This is corresponding to PR splitting as requested in #374.

This adds a complete huge mushroom generator. The old tree generator glowstone used is moved to TreeGenerator but it still the same.

Tree generation from GlowWorld.generateTree() method calls the StructureGrowEvent as it is expected.
#### Todo
- [x] rewrite a better generic tree generator (generates oak tree, small jungle trees and cocoa trees)
- [x] add birch / tall birch tree generator
- [x] add spruce tree generator
#### Test Plugin
- source avalable here: https://gist.github.com/jimmikaelkael/3d3448223c64870bbecd
- jar: https://drive.google.com/file/d/0B0gnjB7PN1h3SlEwZDFMdzdlV1U/view?usp=sharing

### CODE REVIEW

**Comment by gdude2002** [comment]
Looks good. http://bamboo.gserv.me/browse/GLOW-SRV-PR430-92

**Comment by jimmikaelkael** [comment]
Tought I must provide a sample plugin to test the generateTree method with both trees and huge mushrooms...

**Comment by jimmikaelkael** [comment]
@turt2live ready to get reviewed.

**Comment by turt2live** [comment]
@jimmikaelkael This incorrectly reports the number of blocks in a generated tree.

I tested with your test plugin and manually counted the number of blocks in the tree. My results were fairly consistent with the following sample:

```
Expected: 85
Actual: 57
```

Please correct that :)

**Comment by jimmikaelkael** [comment]
This the tree generator doing this: I didn't have modified the code and it's placing some air blocks to build the tree.
EDIT:
I guess I'm good to rewrite the normal Tree generator code.

File: src/main/java/net/glowstone/generator/TreeGenerator.java [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+        // outer leaves
+        for (int j = 0; j < 2; j++) {
+            for (int k = -2; k <= 2; k++) {
+                for (int l = -2; l <= 2; l++) {
+                    delegate.setTypeAndRawData(world, centerX + k, centerY + height - 1 - j, centerZ + l, leaves, data);
+                }
+            }
+        }
+
+        Material air = Material.AIR;
```
Comment (jimmikaelkael) [inline src/main/java/net/glowstone/generator/TreeGenerator.java]: I guess the code below is causing the bad blocks count in StructureGrowEvent.

**Comment by jimmikaelkael** [comment]
@turt2live I think it's ready for a new review.

**Comment by jimmikaelkael** [comment]
Just a note when you'll test this, for the StructureGrowEvent you'll get a discrepancy of 1 block for all trees: this is expected as the block below the trunk is replaced with dirt.
Big mushrooms does not replaces the block below the stem.

EDIT: I have updated the plugin too as I made things a bit more generic: you'll be able to generate small jungle trees and cocoa trees (as they are just a variant of a generic tree).

**Comment by turt2live** [comment]
This looks good, but it's reporting mushrooms backwards. Growing a large red mushroom says it was a brown mushroom and vice-versa.

**Comment by jimmikaelkael** [comment]
I'm really sorry, I inverted the 2 huge mushrooms generation.
This is fixed now.

**Comment by jimmikaelkael** [comment]
Concerning farming PRs, I think this one should be the next to be pulled before #432 can use this to generate the trees on bonemeal interaction with saplings.

**Comment by jimmikaelkael** [comment]
I've rebased it on top of master to ensure all still working, removed those equality checks made with .equals() as it seems SpaceManiac don't like them too much, and I have to admit they are not so much readable.

**Comment by SpaceManiac** [comment]
Merged in 9a5b8227fb7797088d824ff856ec32146564986e.

**Pull request closed without merging**